Modify rule S6597: "WORKDIR" instruction should be used instead of "cd" commands (#3028)
SonarSource/rspec: rules/S6597/docker/rule.adoc

== Why is this an issue?

> In single Dockerfile instructions like `RUN`, `CMD`, and `ENTRYPOINT` multiple commands can be run, including changing directory `cd` command.

Reviewer: ```suggestion
In single Dockerfile instructions like `RUN`, `CMD`, and `ENTRYPOINT` multiple commands can be run, including the `cd` command for changing directories.
```

Using `WORKDIR` instead reduces the complexity of the above instructions and makes them easy to read, understand, troubleshoot, and maintain.

=== What is the potential impact?

The Dockerfile instructions like `RUN` allow users to execute longer scripts. See the following example:

[source,docker]
----
RUN cd /tmp && \
  git clone myrepository.com/MyOrganization/project && \
  cd project && \
  make && \
  cd /app/bin
----

In this example, the first `cd /tmp` command can be replaced by `WORKDIR cd /tmp` put above `RUN`.

> The last `cd /app/bin` command can be replaced too.

Reviewer: I would add the resulting command block for this example as well, that the user can see the "compliant" solution

Those actions will reduce the length of the `RUN` instruction, which makes it easier to read and understand.

> Sometimes, it is hard to avoid the usage of `cd` command, especially in the middle of a long script, but removing them from the beginning and end is straightforward.

Reviewer: ```suggestion
Sometimes, it is hard to avoid the usage of `cd` command, especially in the middle of a long script.
Removing them from the beginning and end of a multi-line is an easy improvement.
```

> Additionally, most commands work well with absolute paths, so changing directories can be avoided very often.

Reviewer: ```suggestion
Additionally, a lot of commands work well with absolute paths, so changing directories can be avoided in most cases.
```

> The `WORKDIR` instruction can be used multiple times in Dockerfile.

Reviewer: ```suggestion
The `WORKDIR` instruction can be used multiple times in a Dockerfile.
```

> It changes the current directory for next instructions (until the following change).

Reviewer: ```suggestion
It changes the current directory for the next instructions and until there is a following change.
```

This approach simplifies understanding of what is a current directory.

The same principles apply to `CMD` and `ENTRYPOINT` instructions.
Following this recommendation provides a clear structure for Dockerfiles, making it easier to maintain.

== How to fix it

> Ensure that all usages of `cd` command are replaced by `WORKDIR` instruction, where possible.

Reviewer: ```suggestion
Where possible, ensure that all usages of `cd` commands are replaced by a `WORKDIR` instruction.
`cd` commands at the beginning or end of multiline commands are a reliable sign, that they can be replaced.
```

> Also, using absolute paths can be considered for commands that accept them.

Reviewer: Why include that in this rule? Isn't that unrelated advice?

Author: It's related, it's another way to avoid `cd` for example in the middle of the script.
